lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java

```diff
@@ -636,16 +636,45 @@ public TopFieldDocs reduce(Collection<TopFieldCollector> collectors) throws IOEx
       query = rewrite(query);
       final Weight weight = createWeight(query, scoreMode, 1);
       final List<Future<C>> topDocsFutures = new ArrayList<>(leafSlices.length);
+
       for (int i = 0; i < leafSlices.length; ++i) {
         final LeafReaderContext[] leaves = leafSlices[i].leaves;
         final C collector = collectors.get(i);
-        topDocsFutures.add(executor.submit(new Callable<C>() {
-          @Override
-          public C call() throws Exception {
+        if (i == leafSlices.length-1) { // execute the last on the caller thread
+          search(Arrays.asList(leaves), weight, collector);
+          topDocsFutures.add(new Future<>() {
+
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning) {
+              return false;
+            }
+
+            @Override
+            public boolean isCancelled() {
+              return false;
+            }
+
+            @Override
+            public boolean isDone() {
+              return true;
+            }
+
+            @Override
+            public C get() {
+              return collector;
+            }
+
+            @Override
+            public C get(long timeout, TimeUnit unit) {
+              return get();
+            }
```

Comment:
Nice Change! I am curious to understand though -- did we see this as an issue, or did we think of this as an optimization? Of course, that is not a blocker to this change

Reply:
well I looked at this code accidentally investigating some idea and I figured it doesn't make sense to just block and wait instead of executing a search so if you are having a single segment you never pay the price of spawning a thread and you don't do any unnecessary work.